pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -629,6 +715,92 @@ func (q *queue) processShadowPartition(ctx context.Context, shadowPart *QueueSha
 	return nil
 }
 
+// normalizeBacklog must be called with exclusive access to the shadow partition
+// NOTE: ideally this is one transaction in a lua script but enqueue_to_backlog is way too much work to
+// utilize
+func (q *queue) normalizeBacklog(ctx context.Context, backlog *QueueBacklog) error {
+	rc := q.primaryQueueShard.RedisClient
+
+	l := logger.StdlibLogger(ctx).With("backlog", backlog)
+
+	// TODO: extend the lease
+
+	var processed int64
+	for {
+		cmd := rc.Client().B().Zrange().Key(rc.kg.BacklogSet(backlog.BacklogID)).Min("-inf").Max("+inf").Byscore().Limit(0, q.backlogNormalizeLimit).Withscores().Build()
+		vals, err := rc.Client().Do(ctx, cmd).AsStrSlice()
+		if err != nil {
+			if rueidis.IsRedisNil(err) {
+				return nil
+			}
+			return fmt.Errorf("could not scan backlog: %w", err)
+		}
+
+		// Done
+		if len(vals) == 0 {
+			return nil
+		}
+
+		itemsToRemove := make([]string, 0)
+		// NOTE: the even idx in the list are the actual item IDs and the odd idx are the scores
+		for i := 0; i < len(vals); i += 2 {
```

Comment:
the retrieval of items here can probably be one lua script to retrieve all in one go instead of making it an iteration and making X amount of request to the queue.

and then just iterate through each queue item and enqueue them again.